test/promql/tests/pipelinerun-cpumins.yaml

```diff
@@ -0,0 +1,136 @@
+# This is the main input for unit testing.
+# Only this file is passed as command line argument.
+
+evaluation_interval: 1m
+
+rule_files:
+  - ../../../prometheus/base/prometheus.rules.yaml
+
+tests:
+  # Test 1.
+  - interval: 1m
+    # Series data.
+    input_series:
+      - series: 'kube_pod_container_resource_limits{resource="memory", namespace="ns1", pod="ns1p1", container="ns1p1c1"}'
+        values: '2x10'
+      - series: 'kube_pod_container_resource_limits{resource="cpu", namespace="ns1", pod="ns1p1", container="ns1p1c1"}'
+        values: '0.5x10'
+      - series: 'kube_pod_labels{label_pipelines_appstudio_openshift_io_type="t1", namespace="ns1", pod="ns1p1"}'
+        values: '1x10'
+      - series: 'container_last_seen{namespace="ns1", pod="ns1p1", container="ns1p1c1"}'
+        values: '200+60x10'
+
+      - series: 'kube_pod_init_container_resource_limits{resource="memory", namespace="ns1", pod="ns1p1", container="ns1p1c2"}'
+        values: '2x10'
+      - series: 'kube_pod_init_container_resource_limits{resource="cpu", namespace="ns1", pod="ns1p1", container="ns1p1c2"}'
+        values: '0.5x10'
+      - series: 'container_last_seen{namespace="ns1", pod="ns1p1", container="ns1p1c2"}'
+        values: '100+60x10'
+
+      - series: 'kube_pod_container_resource_limits{resource="memory", namespace="ns2", pod="ns2p1", container="ns2p1c1"}'
+        values: '1 1 1 4 _ 5 _ 7 9 10 _'
+      - series: 'kube_pod_container_resource_limits{resource="cpu", namespace="ns2", pod="ns2p1", container="ns2p1c1"}'
+        values: '0 5 8 1 _ 10 11 12 12 _ 13'
+      - series: 'kube_pod_labels{label_pipelines_appstudio_openshift_io_type="test", namespace="ns2", pod="ns2p1"}'
+        values: '1x10'
+      - series: 'container_last_seen{namespace="ns2", pod="ns2p1", container="ns2p1c2"}'
+        values: '10 10 20 30 40 0 30 40 60 70 70'
+      - series: 'container_last_seen{namespace="ns2", pod="ns2p1", container="ns2p1c1"}'
+        values: '10 10 20 40 0 40 50 50 60 70 70'
```

Comment:
container_last_seen is a timestamp (unix time in seconds) for: `Last time a container was seen by the exporter` (see here: https://github.com/google/cadvisor/blob/master/docs/storage/prometheus.md#prometheus-container-metrics).

It means that if our measurements interval is 60s, and the container continues to live throughout all measurements, then the value of this metric should grow by 60 on every interval.

It's also possible that it dies somewhere in between measurements, so you may have a single value which is an increase of less than 60 comparing to the previous one, but the measurements afterwards cannot grow with something less than 60.

Let's try to visualize (the actual timestamp is going to look something like `1678956539`, but we're using smaller numbers for convenience):

In this case, c1 is a container that lived throughout the entire measurements, and c2 died just before 60s and somehow resurrected (not sure if that is possible, but anyway) between 180s and 240s.
You see that it has 2 measurements of 57 and afterwards one missing value. I think in reality, it's not going to show the same value twice but instead a value will indeed be missing, but anyhow.

| epoc  | 0 | 60 | 120 | 180 | 240 | 300 | 360 |
| ---  | --- | --- | --- | --- | --- | --- | --- |
| c1_last_seen  | 0 | 60 | 120 | 180 | 240 | 300 | 360 |
| c2_last_seen  | 0 | 57 | 57 | _ | 240 | 300 | 360 |

In other words, any new value appearing will have to be from within the epoc values of the time window it was taken in.

Hope it makes sense...

Reply:
Very nice explanation. 
But why would be a measurement of 57 of 57?
Wouldn't the series look like this?

0 | 57 | _ | _ | 240 | 300 | 360

Reply:
> I think in reality, it's not going to show the same value twice but instead a value will indeed be missing, but anyhow.